author	Pádraig Brady <P@draigBrady.com>	2026-04-28 20:33:10 +0100
committer	Pádraig Brady <P@draigBrady.com>	2026-04-29 13:16:43 +0100
commit	b60a159fdc5bfcf9988d3a4cb6f53abe8ad5d35d (patch)
tree	f7f139bd69bacd6525e190584407d3749897a40c
parent	00cd91288cce5e781dbc750e85ef00a74bc49def (diff)

unexpand: fix heap overflow

* src/unexpand.c (unexpand): Use xinmalloc() to gracefully
handle overflow.  Also use the runtime locale specific MB_CUR_MAX
rather than the worst case MB_LEN_MAX.
* tests/unexpand/mb.sh: Add a test case that fails in a default
glibc build with either MB_CUR_MAX or MB_LEN_MAX.
* NEWS: Mention the bug fix.
Reported by Michał Majchrowicz.

3 files changed, 12 insertions, 1 deletions
Index: coreutils-9.11/NEWS
===================================================================
--- coreutils-9.11.orig/NEWS
+++ coreutils-9.11/NEWS
@@ -4,6 +4,9 @@ GNU coreutils NEWS
 
 ** Bug fixes
 
+  'unexpand -t' no longer overflows a heap buffer, for tab values > SIZE_MAX/16.
+  [bug introduced in coreutils-9.11]
+
   'uniq -w' no longer overruns the read buffer in multibyte locales.
   [bug introduced in coreutils-9.5]
 
Index: coreutils-9.11/src/unexpand.c
===================================================================
--- coreutils-9.11.orig/src/unexpand.c
+++ coreutils-9.11/src/unexpand.c
@@ -131,7 +131,7 @@ unexpand (void)
   /* The worst case is a non-blank character, then one blank, then a
      tab stop, then MAX_COLUMN_WIDTH - 1 blanks, then a non-blank; so
      allocate MAX_COLUMN_WIDTH bytes to store the blanks.  */
-  pending_blank = ximalloc (max_column_width * sizeof (char) * MB_LEN_MAX);
+  pending_blank = xinmalloc (max_column_width, MB_CUR_MAX);
 
   while (true)
     {
Index: coreutils-9.11/tests/unexpand/mb.sh
===================================================================
--- coreutils-9.11.orig/tests/unexpand/mb.sh
+++ coreutils-9.11/tests/unexpand/mb.sh
@@ -17,6 +17,7 @@
 
 . "${srcdir=.}/tests/init.sh"; path_prepend_ ./src
 print_ver_ unexpand printf
+getlimits_
 
 test "$LOCALE_FR_UTF8" != none || skip_ "French UTF-8 locale not available"
 export LC_ALL="$LOCALE_FR_UTF8"
@@ -161,4 +162,11 @@ EOF
 unexpand -a ./in ./in > out || fail=1
 compare exp out > /dev/null 2>&1 || fail=1
 
+# Ensure overflow is handed gracefully
+# coreutils v9.11 induced a buffer overflow with mb_mul=4 (or 16).
+for mb_mul in 4 6; do
+  printf '   \n' | unexpand -t $(expr $SIZE_MAX / $mb_mul + 1) 2>err; ret=$?
+  test "$ret" = 1 || test "$ret" = 0 || { cat err; fail=1; }
+done
+
 Exit $fail
